File: 백준/Gold/17299. 오등큰수/오등큰수.js
const fs = require("fs");
let input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");
const cnt = +input.shift()

input = input[0].split(' ').map(item=>Number(item))

let count = {};
for(let c in input){
    if(!count[input[c]]){
        count[input[c]] = 1;
    }else{
        count[input[c]]++;
    }
}

const stack = []
const result = new Array(cnt).fill(-1)
for(let i=0;i<cnt;i++){
    while(stack.length >0 && count[input[i]] > count[input[stack[stack.length-1]]]){
        result[stack.pop()]=input[i]
    }
    stack.push(i)
}

console.log(result.join(' '))

File: 백준/Gold/17299. 오등큰수/오등큰수.py
import sys
from collections import defaultdict
count=defaultdict(int)
input=sys.stdin.readline
n=int(input())
A=list(map(int,input().split()))
for num in A:
    count[num]+=1
# 1 1 2 3 4 2 1
# 3 3 2 1 1 2 3
# 0 1 2 3 4
stack=[]
ans=[-1]*(n)
for i in range(n):
    while stack and count[A[stack[-1]]]<count[A[i]]:
        ans[stack.pop()]=A[i]

    stack.append(i)

print(' '.join(list(map(str,ans))))
